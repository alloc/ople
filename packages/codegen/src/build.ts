import {
  BinaryExpression,
  BooleanLiteral,
  Expression,
  FunctionDeclaration,
  Identifier,
  ImportDeclaration,
  Node,
  Project,
  SourceFile,
  Statement,
  StringLiteral,
  SyntaxKind,
  ts,
} from 'ts-morph'
import { query as q, FaunaJSON, Expr, ExprVal, FunctionRef } from 'faunadb'
import path from 'path'

export function build() {
  const cwd = process.cwd()
  const project = new Project({
    tsConfigFilePath: path.join(cwd, 'tsconfig.json'),
  })
  const queryDir = project.addDirectoryAtPath(path.join(cwd, 'src/queries'), {
    recursive: true,
  })
  queryDir.getDescendantSourceFiles().forEach(compileQueries)
}

interface QueryDescriptor {
  name: string
  query: string
}

export function compileQueries(file: SourceFile) {
  const locals: { [name: string]: QueryDescriptor } = {}
  const exports: { [name: string]: QueryDescriptor } = {}
  for (const stmt of file.getStatements()) {
    if (Node.isImportDeclaration(stmt)) {
      // TODO: track imported functions
    } else if (Node.isFunctionDeclaration(stmt)) {
      const name = stmt.getName()
      if (!name) continue
      try {
        var lambda = compileLambda(stmt)
      } finally {
        resetContext()
      }
      const query = {
        name,
        query: Expr.toString(lambda, { compact: true }),
      }
      if (stmt.getExportKeyword()) {
        exports[name] = query
      } else {
        locals[name] = query
      }
    } else {
      throw Error('Unsupported top-level statement: ' + stmt.constructor.name)
    }
  }
  return exports
}

function compileLambda(fun: FunctionDeclaration) {
  // TODO: throw if param type is not json-compatible
  const params = fun.getParameters().map(p => p.getName())
  return q.Lambda(params, compileScopedBlock(fun.getStatements(), params))
}

function compileScopedBlock(stmts: Statement[], scope: string[]) {
  pushLocals(scope)
  const block = compileBlock(stmts)
  popLocals()
  return block
}

function compileLambdaWithBlock(stmts: Statement[]) {
  return q.Lambda([], compileBlock(stmts))
}

function compileBlock(stmts: Statement[], needsReturn = true): ExprVal | null {
  const block: ExprVal[] = []
  for (let i = 0; i < stmts.length; i++) {
    let stmt = stmts[i]
    // console.log(stmt.constructor.name)
    if (Node.isReturnStatement(stmt)) {
      const expr = stmt.getExpression()
      block.push(expr ? compileExpression(expr) : null)
      needsReturn = false
      break
    }
    if (Node.isVariableStatement(stmt)) {
      const vars: any = {}
      while (true) {
        for (const decl of stmt.getDeclarations()) {
          const init = decl.getInitializer()
          if (!init) {
            throw Error('Cannot declare a variable without a value')
          }
          const name = decl.getName()
          vars[name] = compileExpression(init)

          // Push to the current scope, so later variables can refer
          // to this variable in their initializer.
          locals.push(name)
        }
        stmt = stmts[i + 1]
        if (Node.isVariableStatement(stmt)) i++
        else break
      }
      block.push(q.Let(vars, compileBlock(stmts.slice(i + 1)) as any))
      needsReturn = false
      break
    }
    if (Node.isExpressionStatement(stmt)) {
      block.push(compileExpression(stmt.getExpression()))
    }
    // if..else
    else if (Node.isIfStatement(stmt)) {
      const thenStmt = stmt.getThenStatement()
      const elseStmt = stmt.getElseStatement()

      let thenBlock = Node.isStatementedNode(thenStmt)
        ? thenStmt.getStatements()
        : null

      let elseBlock = Node.isStatementedNode(elseStmt)
        ? elseStmt.getStatements()
        : null

      const thenReturns = !!thenBlock?.find(Node.isReturnStatement)
      const elseReturns = !!elseBlock?.find(Node.isReturnStatement)

      const shouldAbsorb = thenReturns || elseReturns
      if (shouldAbsorb) {
        const absorbed = stmts.slice(i + 1)
        if (!elseReturns) {
          elseBlock = elseBlock ? elseBlock.concat(absorbed) : absorbed
        } else if (!thenReturns) {
          thenBlock = thenBlock ? thenBlock.concat(absorbed) : absorbed
        }
      }
      block.push(
        q.If(
          compileCondition(stmt.getExpression()),
          thenBlock && compileScopedBlock(thenBlock, []),
          elseBlock && compileScopedBlock(elseBlock, [])
        )
      )
      if (shouldAbsorb) {
        needsReturn = false
        break
      }
    }
  }
  if (needsReturn) {
    block.push(null)
  }
  // console.log(stmts.map(stmt => stmt.constructor.name))
  return block.length > 1 ? q.Do(...block) : block[0] || null
}

function compileCondition(expr: Expression): ExprVal<boolean> {
  const type = checkType(expr)
  const query = compileExpression(expr)
  // TODO: implement CastToBool function
  return type.isBoolean() || type.isBooleanLiteral()
    ? query
    : q.Call('CastToBool', query)
}

function compileIdentifier(node: Identifier) {
  const name = node.getText()
  if (!locals.includes(name)) {
    throw Error('Unknown identifier: ' + JSON.stringify(name))
  }
  return q.Var(name)
}

function compileExpression(expr: Expression): any {
  // console.log(expr.constructor.name)
  if (Node.isIdentifier(expr)) {
    return compileIdentifier(expr)
  }
  if (Node.isParenthesizedExpression(expr)) {
    return compileExpression(expr.getExpression())
  }
  if (Node.isCallExpression(expr)) {
    const callee = expr.getExpression()
    const args = expr.getArguments().map(node => {
      return compileExpression(node as any)
    })

    let func: string | Expr<FunctionRef>
    if (Node.isIdentifier(callee)) {
      func = callee.getText()

      const importDecl = findImportForIdentifier(callee)
      const source = importDecl && importDecl.getModuleSpecifierValue()
      if (source == '@ople/fql') {
        // TODO: map some names to their FQL counterpart
        let name = func[0].toUpperCase() + func.slice(1)
        if (name in q) {
          return (q as any)[name](...args)
        }
      }
    } else {
      const calleeType = checkType(callee)
      if (calleeType.isString()) {
        func = q.Function(compileExpression(callee))
      } else {
        throw Error('Unsupported callee')
      }
    }

    return q.Call(func, ...args)
  }
  if (Node.isPropertyAccessExpression(expr)) {
    if (expr.getQuestionDotTokenNode()) {
      throw SyntaxError('Optional chaining is forbidden')
    }
    const key = expr.getName()
    const left = expr.getExpression()
    const leftType = checkType(left)
    propertyChain.push(key)
    const leftExpr = compileExpression(left)
    propertyChain.pop()
    // console.log(key, leftType.getText())
    if (leftType.isArray()) {
      if (key == 'length') {
        return q.Count(leftExpr)
      }
    } else if (leftType.isString()) {
      if (key == 'length') {
        return q.Length(leftExpr)
      }
    }
    if (Node.isPropertyAccessExpression(left)) {
      return leftExpr
    }
    return q.Select(
      propertyChain.length ? propertyChain.concat(key).reverse() : key,
      leftExpr
    )
  }
  if (Node.isPrefixUnaryExpression(expr)) {
    const op = getOperatorFromToken(expr.getOperatorToken())
    const operand = expr.getOperand()

    const opFn = op && getOperatorFn(op)
    if (!opFn)
      throw Error(
        'Unsupported operator: ' +
          sliceFile(expr.getSourceFile(), expr.getStart(), operand.getStart())
      )

    // Coerce !! to CastToBool call.
    if (op == '!' && startsWith(operand, '!')) {
      return compileCondition(operand.getLastChildOrThrow(Node.isExpression))
    }

    const compileOperand = op == '!' ? compileCondition : compileExpression
    return opFn(compileOperand(operand))
  }
  if (Node.isPostfixUnaryExpression(expr)) {
    const operand = expr.getOperand()
    throw Error(
      'Unsupported operator: ' +
        sliceFile(expr.getSourceFile(), operand.getEnd(), expr.getEnd())
    )
  }
  if (Node.isBinaryExpression(expr)) {
    const op = expr.getOperatorToken().getText()
    const opFn = getOperatorFn(op)
    if (!opFn) {
      throw Error('Unsupported operator: ' + op)
    }
    // TODO: merge subsequent q.Add ops and similar
    return opFn(
      compileExpression(expr.getLeft()),
      compileExpression(expr.getRight())
    )
  }
  if (Node.isRegularExpressionLiteral(expr)) {
    return expr.getLiteralText()
  }
  if (isLiteralNode(expr)) {
    return eval(expr.print())
  }
  return null
}

//
// Utility
//

function checkType(node: Node) {
  return node.getProject().getTypeChecker().getTypeAtLocation(node)
}

function isLiteralNode(node: Node): node is BooleanLiteral | StringLiteral {
  return 'getLiteralValue' in node
}

function findImportForIdentifier(node: Identifier) {
  const [def] = node.getDefinitions()
  return def
    .getNode()
    .getParentWhile(
      (_, child) => !Node.isImportDeclaration(child)
    ) as ImportDeclaration
}

function getOperatorFromToken(op: number) {
  return op == SyntaxKind.ExclamationToken
    ? '!'
    : op == SyntaxKind.TildeToken
    ? '~'
    : null
}

function getOperatorFn(op: string) {
  return op == '>'
    ? q.GT
    : op == '<'
    ? q.LT
    : op == '>='
    ? q.GTE
    : op == '<='
    ? q.LTE
    : op == '=='
    ? q.Equals
    : op == '!'
    ? q.Not
    : op == '+'
    ? q.Add
    : op == '-'
    ? q.Subtract
    : op == '*'
    ? q.Multiply
    : op == '/'
    ? q.Divide
    : op == '%'
    ? q.Modulo
    : op == '&'
    ? q.BitAnd
    : op == '|'
    ? q.BitOr
    : op == '^'
    ? q.BitXor
    : op == '~'
    ? q.BitNot
    : null
}

function sliceFile(file: SourceFile, start: number, end?: number) {
  return file.getText().slice(start, end)
}

function startsWith(node: Node, prefix: string) {
  return (
    sliceFile(
      node.getSourceFile(),
      node.getStart(),
      node.getStart() + prefix.length
    ) == prefix
  )
}

//
// Global context
//

/** Local variable names */
let locals: string[] = []
/** Stack of local scopes */
let scopes: string[][] = []
/** Property access chain */
let propertyChain: string[] = []

function resetContext() {
  locals = []
  scopes = []
  propertyChain = []
}

function pushLocals(names: string[]) {
  scopes.push(locals)
  locals = locals.concat(names)
}

function popLocals() {
  locals = scopes.pop() || []
}
