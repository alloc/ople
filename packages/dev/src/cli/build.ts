import { app } from './app'

app.command('build', 'Build the production server').action(() => {
  // TODO
  //  - rollup
  //  - adapter for firebase functions
  //  - deploy script for firebase/pushpin
})
