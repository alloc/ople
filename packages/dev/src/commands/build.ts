import { app } from '../core'

app.command('build', 'Build the production server').action(() => {
  // TODO
  //  - rollup
  //  - adapter for firebase functions
  //  - deploy script for firebase/pushpin
})
