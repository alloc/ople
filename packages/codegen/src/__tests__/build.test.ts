import dedent from 'dedent'
import { Project } from 'ts-morph'
import { compileQueries } from '../build'

const project = new Project()
const makeFile = (code: string) =>
  project.createSourceFile(
    project.getSourceFiles().length + '.ts',
    dedent(code)
  )

afterEach(() => {
  project.getSourceFiles().forEach(file => project.removeSourceFile(file))
})

describe('queries', () => {
  const compileFile = (code: string) => compileQueries(makeFile(code))

  test('export function', () => {
    expect(compileFile(`export function foo() {}`)).toMatchSnapshot()
  })

  test('unknown identifier', () => {
    const file = makeFile(`
      export function foo() {
        return lol
      }
    `)

    expect(() => compileQueries(file)).toThrowErrorMatchingSnapshot()
  })

  test('literal values', () => {
    const file = makeFile(`
      export function foo() {
        true
        false
        'string'
        null
        123
      }
    `)

    expect(compileQueries(file)).toMatchSnapshot()
  })

  test('declare variable', () => {
    const file = makeFile(`
      export function foo() {
        let a = 1
      }
    `)

    expect(compileQueries(file)).toMatchSnapshot()
  })

  test('early return', () => {
    const file = makeFile(`
      export function foo() {
        return 1
        return 2
      }
    `)

    expect(compileQueries(file)).toMatchSnapshot()
  })

  describe('if blocks', () => {
    test('empty then', () => {
      const file = makeFile(`
        export function foo() {
          if (true) {}
        }
      `)

      expect(compileQueries(file)).toMatchSnapshot()
    })

    test('then return', () => {
      const file = makeFile(`
        export function foo() {
          if (true) {
            return 1
          }
          return 2
        }
      `)

      expect(compileQueries(file)).toMatchSnapshot()
    })

    test('else return', () => {
      const file = makeFile(`
        export function foo() {
          if (true) {}
          else {
            return 1
          }
          return 2
        }
      `)

      expect(compileQueries(file)).toMatchSnapshot()
    })

    test('nested return', () => {
      const file = makeFile(`
        export function foo(doc: any) {
          if (true) {
            if (true) {
              return 1
            }
            doc.data.foo = 1
          }
          return 2
        }
      `)

      expect(compileQueries(file)).toMatchInlineSnapshot(`null`)
    })

    test('non-boolean condition', () => {
      const file = makeFile(`
        export function foo(arg: string) {
          if (arg) {}
        }
      `)

      expect(compileQueries(file)).toMatchSnapshot()
    })
  })

  describe('operators', () => {
    const compileExpression = (expr: string) =>
      compileFile(`
        export function foo(a: number, b: number) {
          return ${expr}
        }
      `)

    test('algebraic operators', () => {
      expect([
        compileExpression('a + b'),
        compileExpression('a - b'),
        compileExpression('a * b'),
        compileExpression('a / b'),
        compileExpression('a % b'),
      ]).toMatchSnapshot()
    })

    test('bitwise operators', () => {
      expect([
        compileExpression('a & b'),
        compileExpression('a | b'),
        compileExpression('a ^ b'),
        compileExpression('~a'),
      ]).toMatchSnapshot()
    })

    test('comparison operators', () => {
      expect([
        compileExpression('a > b'),
        compileExpression('a >= b'),
        compileExpression('a < b'),
        compileExpression('a <= b'),
        compileExpression('a == b'),
      ]).toMatchSnapshot()
    })

    test('not operator', () => {
      expect([
        compileExpression('!a'),
        compileExpression('!!a'),
      ]).toMatchInlineSnapshot(`null`)
    })

    test('strict equals', () => {
      expect(() => compileExpression('a === b')).toThrowErrorMatchingSnapshot()
    })
  })

  describe('strings', () => {
    test('length', () => {
      const result = compileFile(`
        export function foo(arg: string) {
          return arg.length
        }
      `)

      expect(result).toMatchSnapshot()
    })
  })

  describe('objects', () => {
    test('single property', () => {
      const result = compileFile(`
        export function foo(obj: {a: any}) {
          return obj.a
        }
      `)

      expect(result).toMatchSnapshot()
    })

    test('nested property', () => {
      const result = compileFile(`
        export function foo(obj: {a: {b:any}}) {
          return obj.a.b
        }
      `)

      expect(result).toMatchSnapshot()
    })
  })

  describe('arrays', () => {
    test('length', () => {
      const result = compileFile(`
        export function foo(arg: any[]) {
          return arg.length
        }
      `)

      expect(result).toMatchSnapshot()
    })
  })

  describe('documents', () => {})

  describe('collections', () => {})

  describe('stdlib', () => {
    test('toArray', () => {})
    test('isArray', () => {})
  })
})
