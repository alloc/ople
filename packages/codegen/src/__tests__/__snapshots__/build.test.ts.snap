// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries arrays length 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([\\"arg\\"],Count(Var(\\"arg\\")))",
  },
}
`;

exports[`queries declare variable 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],Let([{a:1}],null))",
  },
}
`;

exports[`queries early return 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],1)",
  },
}
`;

exports[`queries export function 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],null)",
  },
}
`;

exports[`queries if blocks else return 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],If(true,2,1))",
  },
}
`;

exports[`queries if blocks empty then 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],Do(If(true,null,null),null))",
  },
}
`;

exports[`queries if blocks non-boolean condition 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([\\"arg\\"],Do(If(Call(\\"CastToBool\\",Var(\\"arg\\")),null,null),null))",
  },
}
`;

exports[`queries if blocks then return 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],If(true,1,2))",
  },
}
`;

exports[`queries literal values 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([],Do(true,false,\\"string\\",null,123,null))",
  },
}
`;

exports[`queries objects nested property 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([\\"obj\\"],Select([\\"a\\",\\"b\\"],Var(\\"obj\\")))",
  },
}
`;

exports[`queries objects single property 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([\\"obj\\"],Select(\\"a\\",Var(\\"obj\\")))",
  },
}
`;

exports[`queries operators algebraic operators 1`] = `
Array [
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Add(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Subtract(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Multiply(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Divide(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Modulo(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
]
`;

exports[`queries operators bitwise operators 1`] = `
Array [
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Bitand([Var(\\"a\\"),Var(\\"b\\")]))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Bitor([Var(\\"a\\"),Var(\\"b\\")]))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Bitxor([Var(\\"a\\"),Var(\\"b\\")]))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Bitnot(Var(\\"a\\")))",
    },
  },
]
`;

exports[`queries operators comparison operators 1`] = `
Array [
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],GT(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],GTE(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],LT(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],LTE(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Equals(Var(\\"a\\"),Var(\\"b\\")))",
    },
  },
]
`;

exports[`queries operators not operator 1`] = `
Array [
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Not(Call(\\"CastToBool\\",Var(\\"a\\"))))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Not(GT(Var(\\"a\\"),0)))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Not(Call(\\"CastToBool\\",Add(Var(\\"a\\"),Var(\\"b\\")))))",
    },
  },
  Object {
    "foo": Object {
      "name": "foo",
      "query": "Lambda([\\"a\\",\\"b\\"],Call(\\"CastToBool\\",Var(\\"a\\")))",
    },
  },
]
`;

exports[`queries operators strict equals 1`] = `"Unsupported operator: ==="`;

exports[`queries strings length 1`] = `
Object {
  "foo": Object {
    "name": "foo",
    "query": "Lambda([\\"arg\\"],Length(Var(\\"arg\\")))",
  },
}
`;

exports[`queries unknown identifier 1`] = `"Unknown identifier: \\"lol\\""`;
